lib/auth/join_gcp.go

```diff
@@ -0,0 +1,97 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auth
+
+import (
+	"context"
+	"strings"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/exp/slices"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/gcp"
+)
+
+type gcpIDTokenValidator interface {
+	Validate(ctx context.Context, token string) (*gcp.IDTokenClaims, error)
+}
+
+func (a *Server) checkGCPJoinRequest(ctx context.Context, req *types.RegisterUsingTokenRequest) error {
+	if req.IDToken == "" {
+		return trace.BadParameter("IDToken not provided for GCP join request")
+	}
+	pt, err := a.GetToken(ctx, req.Token)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	token, ok := pt.(*types.ProvisionTokenV2)
+	if !ok {
+		return trace.BadParameter("gcp join method only supports ProvisionTokenV2, '%T' was provided", pt)
+	}
+
+	claims, err := a.gcpIDTokenValidator.Validate(ctx, req.IDToken)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	log.WithFields(logrus.Fields{
+		"claims": claims,
+		"token":  pt.GetName(),
+	}).Info("GCP VM trying to join cluster")
+
+	return trace.Wrap(checkGCPAllowRules(token, claims))
+}
+
+func checkGCPAllowRules(token *types.ProvisionTokenV2, claims *gcp.IDTokenClaims) error {
+	compute := claims.Google.ComputeEngine
+	// If a single rule passes, accept the IDToken.
+	for _, rule := range token.Spec.GCP.Allow {
+		if len(rule.ProjectIDs) > 0 && !slices.Contains(rule.ProjectIDs, compute.ProjectID) {
+			continue
+		}
+		if len(rule.Locations) > 0 && !slices.ContainsFunc(rule.Locations, func(location string) bool {
+			return isGCPZoneInLocation(location, compute.Zone)
+		}) {
+			continue
+		}
+		if len(rule.ServiceAccounts) > 0 && !slices.Contains(rule.ServiceAccounts, claims.Email) {
+			continue
+		}
+
+		// All provided rules met.
+		return nil
+	}
+
+	return trace.AccessDenied("id token claims did not match any allow rules")
+}
+
+// isGCPZoneInLocation checks if a zone belongs to a location, which can be
+// either a zone or region.
+func isGCPZoneInLocation(location, zone string) bool {
+	if zone == location {
+		return true
+	}
+	locationParts := strings.Split(location, "-")
+	// location is not a region.
+	if len(locationParts) != 2 {
+		return false
+	}
+	return strings.HasPrefix(zone, locationParts[0]+"-"+locationParts[1])
```

Comment:
Is there a reason to build back `location` from `locationParts` instead of using `location` directly?

Also, I think this check is not strong enough and we must split `zone` in parts and check equality.

e.g.
```go
// As a use I want to restrict the token to the region europe-west1
location := "europe-west1"
zone := "europe-west12-a"

// locationParts = ["europe", "west1"]
// HasPrefix returns true while the token comes from another region
```

I suppose this was made to allow rules like `europe-west` but this feels like a potential vulnerability. If we want to support such patterns, maybe we could use a wildcard-like character to know if we should use prefix-matching or exact matching. e.g. `europe-west12` would match `europe-west*` but not `europe-west1`.

Reply:
Having the user list all the existing regions is also a valid UX IMO